### PR TITLE
Reimplement std's thread parker on top of events on SGX

### PR DESCRIPTION
**Pull request by joboet**
Mutex and Condvar are being replaced by more efficient implementations, which need thread parking themselves (see #93740). Therefore, the generic `Parker` needs to be replaced on all platforms where the new lock implementation will be used.

SGX enclaves have a per-thread event state, which allows waiting for and setting specific bits. This is already used by the current mutex implementation. The thread parker can however be much more efficient, as it only needs to store the `TCS` address of one thread. This address is stored in a state variable, which can also be set to indicate the thread was already notified.

`park_timeout` does not guard against spurious wakeups like the current condition variable does. This is allowed by the API of `Parker`, and I think it is better to let users handle these wakeups themselves as the guarding is quite expensive and might not be necessary.

@jethrogb as you wrote the initial SGX support for `std`, I assume you are the target maintainer? Could you help me test this, please? Lacking a x86_64 chip, I can't run SGX.

### CODE REVIEW

**Comment by jethrogb** [comment]
Cool stuff. It will take use some time to review this to ensure security. cc @mzohreva @raoulstrackx

**Comment by Mark-Simulacrum** [comment]
Once you have reviewed, please re-assign back to me or someone else for the final r+

File: library/std/src/sys/sgx/thread_parker.rs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    pub unsafe fn new(parker: *mut Parker) {
+        unsafe { parker.write(Parker::new_internal()) }
+    }
+
+    pub(super) fn new_internal() -> Parker {
+        Parker { state: AtomicPtr::new(EMPTY) }
+    }
+
+    // This implementation doesn't require `unsafe` and `Pin`, but other implementations do.
+    pub unsafe fn park(self: Pin<&Self>) {
```
Comment (raoulstrackx) [inline library/std/src/sys/sgx/thread_parker.rs:32]: Can you provide documentation of what properties the `park` and `unpark` functions are expected to provide?

Comment (joboet) [reply library/std/src/sys/sgx/thread_parker.rs:32]: These functions are exposed publicly as [`thread::park`](https://doc.rust-lang.org/nightly/std/thread/fn.park.html) and [`Thread::unpark`](https://doc.rust-lang.org/nightly/std/thread/struct.Thread.html#method.unpark), respectively, where they are also documented.

File: library/std/src/sys/sgx/thread_parker.rs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        let tcs = thread::current().as_ptr();
+
+        if self.state.load(Acquire) != NOTIFIED {
```
[the remark this answers is not in the document]
Comment (joboet) [reply library/std/src/sys/sgx/thread_parker.rs:35]: Exactly. I will add a comment explaining this...

File: library/std/src/sys/sgx/thread_parker.rs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+use super::abi::{thread, usercalls};
+use crate::io::ErrorKind;
+use crate::pin::Pin;
+use crate::ptr::{self, NonNull};
+use crate::sync::atomic::AtomicPtr;
+use crate::sync::atomic::Ordering::{Acquire, Relaxed, Release};
+use crate::time::Duration;
+use fortanix_sgx_abi::{EV_UNPARK, WAIT_INDEFINITE};
+
+const EMPTY: *mut u8 = ptr::invalid_mut(0);
```
[the remark this answers is not in the document]
Comment (joboet) [reply library/std/src/sys/sgx/thread_parker.rs:12]: Of course. `Tcs` being a `NonNull` pointer lead me to believe that assumption was correct.

File: library/std/src/sys/sgx/thread_parker.rs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+                return;
+            }
+        }
+
+        // The token was already read with `acquire` ordering, this can be a store.
+        self.state.store(EMPTY, Relaxed);
+    }
+
+    // This implementation doesn't require `Pin`, but other implementations do.
+    pub fn unpark(self: Pin<&Self>) {
```
[the remark this answers is not in the document]
Comment (joboet) [reply library/std/src/sys/sgx/thread_parker.rs:93]: I've now explained that situation in the comments. I hope that resolves this?

File: library/std/src/sys/sgx/thread_parker.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jethrogb) [reply library/std/src/sys/sgx/thread_parker.rs]: @raoulstrackx the TCS position is covered by MRENCLAVE so we can easily define that there will never be TCS at address 0?

[the remark this answers is not in the document]
Comment (raoulstrackx) [reply library/std/src/sys/sgx/thread_parker.rs]: Yes we can, but as far as I know, we don't specify anywhere limitations on how the enclave should be laid out in memory. So either we should specify it somewhere, or we can't assume that the tool that does the lay out the enclave doesn't put a TCS struct as the first page of the enclave.

[the remark this answers is not in the document]
Comment (jethrogb) [reply library/std/src/sys/sgx/thread_parker.rs]: I'd say this is implied by https://github.com/fortanix/rust-sgx/blob/master/intel-sgx/fortanix-sgx-abi/src/lib.rs#L446 but I'd accept a documentation PR.

[the remark this answers is not in the document]
Comment (raoulstrackx) [reply library/std/src/sys/sgx/thread_parker.rs]: For future reference: https://github.com/fortanix/rust-sgx/pull/408

File: library/std/src/sys/sgx/thread_parker.rs (hunk 1 above)
Comment (raoulstrackx) [reply library/std/src/sys/sgx/thread_parker.rs:38]: Sorry, I lost track of this PR. I meant security properties, but I guess having them in `Thread::park` and `Thread::unpark` suffices.

**Pull request merged**